### ISSUE TITLE
Type narrowing for "in" is only working for tuples

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#3283]
```python
from typing import Literal

var_type = Literal["foo", "bar", "bin"]

def some_func() -> var_type:
    ...

var = some_func()

if var == "foo" or var == "bar":
    reveal_type(var)

if var in ("foo", "bar"):
    reveal_type(var)

if var in ["foo", "bar"]:
    reveal_type(var)

if var in {"foo", "bar"}:
    reveal_type(var)
```

The following produces:

```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx.py
  /tmp/xxx.py:13:17 - information: Type of "var" is "Literal['foo', 'bar']"
  /tmp/xxx.py:16:17 - information: Type of "var" is "Literal['foo', 'bar']"
  /tmp/xxx.py:19:17 - information: Type of "var" is "Literal['foo', 'bar', 'bin']"
  /tmp/xxx.py:22:17 - information: Type of "var" is "Literal['foo', 'bar', 'bin']"
0 errors, 0 warnings, 4 informations 
Completed in 0.598sec
```

Only the `or` and the `in tuple` is narrowing the types correctly, while the `in list` (most common type used for this kind of narrowing) and `in set` is not.

I noticed that with `Literal`, but might be affecting other types as well.

**Comment by erictraut** [comment]
The behavior is correct. The inferred type of the tuple expression `("foo", "bar")` is `tuple[Literal["foo"], Literal["bar"]]`. By contrast, the inferred type of the list expression `["foo", "bar"]` is `list[str]`, and the inferred type of the set expression `{"foo", "bar"}` is `set[str]`.

**Comment by bellini666** [comment]
Hey @erictraut ,

Although I understand the reason of the inferred type being that, the use case here is not correct.

I mean, a tuple is something that the positions have meanings, while a list is a collection of something. Thus, although checking `if var in ("foo", "bar")` works due to the sequence nature of the tuple, it is not semantically correct and also very strange to do, which is why using `if var in ["foo", "bar"]` is used mostly.

Also, `if var in ["foo", "bar"]` is the recommended pythonic way of doing `if var == "foo" or var == "bar"`.

Because of those reasons, don't you think this kind of narrowing should follow its own rules?

**Comment by erictraut** [comment]
List and set expressions do not retain literal types in the Python type system because they are mutable. By contrast, tuples can retain literal types because they are immutable. Type narrowing logic needs to live within these constraints.

If you want type narrowing to work in this case, you should use one of the two supported patterns.

**Comment by bellini666** [comment]
Yes, I know that. My suggestion was to consider that specific context, where the list/set is being created just for the check and not being stored in a variable example.

I can understand if you don't want to support this right now, but I'd like to ask for you to try to reconsider this in the future. Like I said previously, it is a very common and recommended pattern. Forcing the use of tuples is unnatural and leads to people misusing them...

**Comment by erictraut** [comment]
It's unlikely that this will change.

I think you need to change the way you're thinking about static type checking. The purpose of a static type checker is to enable you to write robust, maintainable, bug-free code. The purpose is _not_ to support every possible coding pattern under all circumstances. I encourage you to make a mental shift if you want to take advantage of static type checking. There are some existing code patterns that appear in Python that are not compatible with static type checking, and these will evolve as more people embrace static type checking and its benefits.
